internal/backend/local/test.go

```diff
@@ -628,7 +610,46 @@ func (runner *TestFileRunner) run(run *moduletest.Run, file *moduletest.File, st
 	// our prior run outputs so future run blocks can access it.
 	runner.EvalContext.SetOutput(run, outputVals)
 
-	return updated, true
+	// Only update the most recent run and state if the state was
+	// actually updated by this change. We want to use the run that
+	// most recently updated the tracked state as the cleanup
+	// configuration.
+	runner.EvalContext.SetFileState(key, &graph.TestFileState{
+		Run:   run,
+		State: updated,
+	})
+}
+
+// checkForMissingExpectedFailures checks for missing expected failures in the diagnostics.
+// It updates the run status based on the presence of errors or missing expected failures.
+func (runner *TestFileRunner) checkForMissingExpectedFailures(run *moduletest.Run, diags tfdiags.Diagnostics) (failOrErr bool) {
+	// If there are errors in the diagnostics, update the run status to error.
+	if diags.HasErrors() {
+		run.Status = moduletest.Error
+	}
+
+	// Retrieve diagnostics that are either unrelated to expected failures or report missing expected failures.
+	unexpectedDiags := run.ValidateExpectedFailures(diags)
+	var nonFailureDiags tfdiags.Diagnostics
+	for _, diag := range unexpectedDiags {
+		switch {
+		// If any diagnostic indicates a missing expected failure, update the run status to fail.
+		case diag.Description().Summary == moduletest.MissingFailureSummary:
+			diag.ExtraInfo()
+			run.Status = moduletest.Fail
```

Comment:
```suggestion
			run.Status = run.Status.Merge(moduletest.Fail)
```

This just makes sure that if it has already errored earlier for some reason we don't overwrite that with a Fail at this point. Generally, you should do this everywhere but when setting Error it doesn't matter so much as that has the highest priority.